### PR TITLE
Create memory context awareness

### PR DESCRIPTION
**Pull request by bluejekyll**

### CODE REVIEW

**Comment by intgr** [comment]
Is this from Discord still an up to date explanation for the general approach in this PR?

@bluejekyll Last Wednesday at 18:56

> Here's my current plan, all data coming into the function from PG will be owned by the current memory context when the Rust function is called. This is the FUNC_ALLOCATOR global static. (A better name for this would be CURRENT_PG_MEMORY_CONTEXT). This provides a `'static` lifetime on function parameters (`PgDatum<'mc>` being passed into the function will be `PgDatum<'static>`). Any data returned from the function must be the same lifetime as the passed `PgAllocator` passed into the function (this will be an optional binding and the pg-extern attribute needs to be updated for this). I'm going go change the current `TryFromPgDatum` to always return `PgAllocated`, and then for base types we'll provide a conversion from `PgAllocated` to Rust types (current impl has some issues there). I plan to add some methods to the PgAllocator, specifically, it should be possible to set the current memory context (and roll it back) in PG when creating a new memory context, and before calling into PG, we can use RAII to reset the memory context after. In addition, if we may want the top-level function to be able to change the MemoryContext for returned data (or maybe not?) and for that we'd need a method to swap out the FUNC_ALLOCATOR. Anyway, I think this melds the MemoryContext with lifetimes in Rust in a safe way, and when combined with the `PgAllocated` type, prevents unsafe sharing of data across threads or outside the lifetime of the owning MemoryContext.

**Comment by bluejekyll** [comment]
Mostly. I think I’ve managed to drop the requirement for a global static (FUNC_ALLOCATOR), so that won’t be part of the picture.

I’m still working on some of the details around how we execute functions in the proper context. I made a little progress on that last night.

File: examples/fdw-rw/src/lib.rs [hunk 1]
```diff
@@ -4,13 +4,14 @@
 // http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
 // http://opensource.org/licenses/MIT>, at your option. This file may not be
 // copied, modified, or distributed except according to those terms.
+#![cfg(fdw)]
```
Comment (bluejekyll) [inline examples/fdw-rw/src/lib.rs]: remove this before checkin

File: integration-tests/tests/fdw.rs [hunk 2]
```diff
@@ -1,8 +1,11 @@
+#![cfg(fdw)]
+
 extern crate integration_tests;
 
 use integration_tests::*;
 
 #[test]
+#[ignore]
```
Comment (bluejekyll) [inline integration-tests/tests/fdw.rs]: remove before checkin

File: integration-tests/tests/fdw.rs [hunk 3]
```diff
@@ -1,8 +1,11 @@
+#![cfg(fdw)]
```
Comment (bluejekyll) [inline integration-tests/tests/fdw.rs]: remove before checkin

File: integration-tests/tests/strings.rs [hunk 4]
```diff
@@ -14,3 +14,11 @@ fn test_concat_rs() {
         assert_eq!(&col, "ab");
     });
 }
+
+#[test]
+fn test_text_rs() {
+    test_in_db("strings", |conn| {
+        let result = conn.query("SELECT text_rs('hello world!')", &[]).expect("query failed");
+        assert_eq!(result.len(), 1);
```
Comment (bluejekyll) [inline integration-tests/tests/strings.rs]: add a test to verify the text returned.

File: pg-extend/src/native/varlena.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    pub(crate) fn as_slice(&self) -> &[std::os::raw::c_char] {
+        use VarLenA::*;
+
+        let len = self.len();
+
+        unsafe {
+            match self {
+                VarAtt4b(varlena) => varlena.va_data.as_slice(len),
+                VarAtt1b(varlena) => varlena.va_data.as_slice(len),
+                _ => unimplemented!("this VarLenA type not yet supported"),
```
Comment (bluejekyll) [inline pg-extend/src/native/varlena.rs]: add `{:?}` for more debug info

File: pg-extend/src/native/varlena.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                ((varlena.va_header >> 2) & 0x3FFFFFFF) as usize - Self::size_of(&varlena.va_header)
+            }
+            // #define VARSIZE_1B(PTR) \
+            // ((((varattrib_1b *) (PTR))->va_header >> 1) & 0x7F)
+            VarAtt1b(varlena) => {
+                ((varlena.va_header >> 1) & 0x7F) as usize - Self::size_of(&varlena.va_header)
+            }
+            // #define VARTAG_1B_E(PTR) \
+            // (((varattrib_1b_e *) (PTR))->va_tag)
+            _ => unimplemented!("this VarLenA type not yet supported"),
```
Comment (bluejekyll) [inline pg-extend/src/native/varlena.rs]: add `{:?}` for more debug info

File: pg-extend/src/native/varlena.rs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    VarAtt4bU,
+    VarAtt4bC,
+    VarAtt1b(&'a pg_sys::varattrib_1b),
+    VarAtt1bE,
+    VarAttNotPadByte,
+}
+
+impl<'a> VarLenA<'a> {
+    /// See postgres.h
+    pub(crate) unsafe fn from_varlena(varlena: &'a pg_sys::varlena) -> VarLenA<'a> {
```
Comment (bluejekyll) [inline pg-extend/src/native/varlena.rs]: This is going to have forward compatibility issues.

**Pull request merged**